### PR TITLE
Read Only Root Filesystem compatibility

### PR DESCRIPTION
**Pull request by GauntletWizard**
**Description of the change**

Makes the built image compatible with readonly root fs, i.e. `docker run --read-only`

**Benefits**

This change is intended to make it easier and simpler for developers running on kubernetes with a Pod Security Policy that forces read-only root filesystems to run the bitnami image. It makes clearer which directories the image will write to.

**Possible drawbacks**

There's a number of items in the VOLUME stanza, which can leave some clutter around if you're not careful with your container cleanup. 

**Applicable issues**

None

**Additional information**

### CODE REVIEW

File: 4.0/debian-9/Dockerfile [hunk 1]
```diff
@@ -24,8 +24,13 @@ ENV ALLOW_EMPTY_PASSWORD="no" \
     REDIS_PASSWORD="" \
     REDIS_REPLICATION_MODE=""
 
+RUN mv /opt/bitnami/redis/etc /opt/bitnami/redis/etc-default
```
Comment (juan131) [inline 4.0/debian-9/Dockerfile]: We can modify the "Redis package" so it contains the default configuration file at **/opt/bitnami/redis/etc-default**. This way we don't need to add an unnecessary extra lawyer in the Dockerfile and keep the changes you suggested in the library `libredis.sh`

File: 4.0/debian-9/Dockerfile [hunk 2]
```diff
@@ -24,8 +24,13 @@ ENV ALLOW_EMPTY_PASSWORD="no" \
...
+RUN install -d -o 1001 -g 1001 /bitnami/redis/data /opt/bitnami/redis/etc /opt/bitnami/redis/logs /opt/bitnami/redis/tmp /tmp
+
 EXPOSE 6379
 
 USER 1001
+VOLUME [ "/bitnami/redis/data", "/opt/bitnami/redis/etc", "/opt/bitnami/redis/logs", "/opt/bitnami/redis/tmp", "/tmp"]
+WORKDIR "/tmp"
```
Comment (juan131) [inline 4.0/debian-9/Dockerfile]: Why do you want to use this working directory?

File: 4.0/debian-9/Dockerfile [hunk 3]
```diff
@@ -24,8 +24,13 @@ ENV ALLOW_EMPTY_PASSWORD="no" \
     REDIS_PASSWORD="" \
     REDIS_REPLICATION_MODE=""
 
+RUN mv /opt/bitnami/redis/etc /opt/bitnami/redis/etc-default
+RUN install -d -o 1001 -g 1001 /bitnami/redis/data /opt/bitnami/redis/etc /opt/bitnami/redis/logs /opt/bitnami/redis/tmp /tmp
```
Comment (juan131) [inline 4.0/debian-9/Dockerfile]: Please move whatever changes you want to do on these directories to `prepares.sh` file.

File: 4.0/debian-9/Dockerfile [hunk 4]
```diff
@@ -24,8 +24,13 @@ ENV ALLOW_EMPTY_PASSWORD="no" \
...
+
 EXPOSE 6379
 
 USER 1001
+VOLUME [ "/bitnami/redis/data", "/opt/bitnami/redis/etc", "/opt/bitnami/redis/logs", "/opt/bitnami/redis/tmp", "/tmp"]
```
Comment (juan131) [inline 4.0/debian-9/Dockerfile]: What is the main advantage of creating the mount point in advance?

File: 4.0/debian-9/Dockerfile (hunk 2 above)
Comment (GauntletWizard) [reply 4.0/debian-9/Dockerfile]: Somewhat of an artifact. Bash attempts to create a temporary file when evaluating heredocs. Specifically, what I ran into was the invocation of `eval $(redis_env)` fails if the working directory is not writable.

File: 4.0/debian-9/Dockerfile (hunk 4 above)
Comment (GauntletWizard) [reply 4.0/debian-9/Dockerfile]: The docker runtime will automatically populate listed volumes with volumes scoped to the lifetime of the container; i.e. this makes it "just work" by making these directories (and only these directories) writable even when using the `--read-only` flag.

(Other container runtimes differ in behavior, but even if it doesn't "just work" it serves as documentation of what paths need mounts to be writable.)

**Comment by GauntletWizard** [comment]
Ping. I've updated the branch to comply with style, and answered your comments above. Let me know if there's any more info I can provide or if you've got other reservations.

File: 4.0/debian-9/Dockerfile (hunk 4 above)
Comment (juan131) [reply 4.0/debian-9/Dockerfile]: Thanks for the great explanation! I didn't know about these advantages.

File: 4.0/debian-9/Dockerfile (hunk 2 above)
Comment (juan131) [reply 4.0/debian-9/Dockerfile]: 👍

File: 4.0/debian-9/rootfs/libredis.sh [hunk 5]
```diff
@@ -308,19 +309,15 @@ redis_initialize() {
     info "Initializing Redis..."
 
     # User injected custom configuration
-    if [[ -e "$REDIS_BASEDIR/etc/redis.conf" ]]; then
-        if [[ -e "$REDIS_BASEDIR/etc/redis-default.conf" ]]; then
-            rm "${REDIS_BASEDIR}/etc/redis-default.conf"
-        fi
-    else
+    if [[ ! -e "$REDIS_BASEDIR/etc/redis.conf" ]]; then
```
Comment (juan131) [inline 4.0/debian-9/rootfs/libredis.sh]: Now you defined the env. variable `REDIS_CONFDIR`, I'd use it here.

**Comment by juan131** [comment]
Please note we can't merge the PR until we modify the "Redis package" so it contains the default configuration file at `/opt/bitnami/redis/etc-default`

**Comment by juan131** [comment]
@tompizmor I'd like you to review this PR

**Comment by GauntletWizard** [comment]
@tompizmor Ping?

**Comment by tompizmor** [comment]
@GauntletWizard Sorry I missed the PR. I built the image and did some tests. I had to modify the redis bitnami package to include the conf in `etc-default` as @juan131 said. For testing purposes, I pushed the image to `tompizmor/redis:rofs`

I found the next issue testing the image in kubernetes with the stable/redis chart with `readOnlyRootFilesystem` enabled for the redis container. 

On master pod:

```
1:M 23 Jan 15:52:44.902 * Slave 10.32.9.11:6379 asks for synchronization
1:M 23 Jan 15:52:44.902 * Full resync requested by slave 10.32.9.11:6379
1:M 23 Jan 15:52:44.902 * Starting BGSAVE for SYNC with target: disk
1:M 23 Jan 15:52:44.903 * Background saving started by pid 325
325:C 23 Jan 15:52:44.904 # Failed opening the RDB file dump.rdb (in server root dir /data) for saving: Permission denied
1:M 23 Jan 15:52:44.904 # Background saving error
1:M 23 Jan 15:52:44.904 # Connection with slave 10.32.9.11:6379 lost.
1:M 23 Jan 15:52:44.905 # SYNC failed. BGSAVE child returned an error
```

On slave pod:

```
1:S 23 Jan 15:52:37.871 * Connecting to MASTER redis-master:6379
1:S 23 Jan 15:52:37.873 * MASTER <-> SLAVE sync started
1:S 23 Jan 15:52:37.873 * Non blocking connect for SYNC fired the event.
1:S 23 Jan 15:52:37.873 * Master replied to PING, replication can continue...
1:S 23 Jan 15:52:37.873 * Partial resynchronization not possible (no cached master)
1:S 23 Jan 15:52:37.874 * Full resync from master: eb2327e49fc6c49fc0c9a728f69c0a101cf78151:0
1:S 23 Jan 15:52:37.887 # I/O error reading bulk count from MASTER: Resource temporarily unavailable
```

So it seems there is some permissions misconfigured in the data folder.

**Pull request closed without merging**